# Introduction to SQL Using Python: Aggregating Data with the GROUP BY Statement

In my last blog, I discussed how to filer data using SQL with the __WHERE__ statement. This blog is a tutorial on how to aggregate data using the __GROUP BY__ statement.

In [1]:
# Import necessary libraries

import sqlite3
import pandas as pd

In [2]:
# Connect to database
conn = sqlite3.connect('''database.sqlite''')

# Create cursor object
cur = conn.cursor()

In [ ]:
cur.execute('''Enter SQL query here;''') # Runs SQL query
data = pd.DataFrame(cur.fetchall()) # Converts SQL query results into dataframe format
data.columns = [x[0] for x in cur.description] # Labels the columns of the dataframe
data # View SQL results dataframe

In [7]:
# View Teams dataframe

cur.execute('''SELECT * 
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Season        TeamName KaderHome AvgAgeHome ForeignPlayersHome  \
0     2017   Bayern Munich        27         26                 15   
1     2017        Dortmund        33         25                 18   
2     2017      Leverkusen        31         24                 15   
3     2017      RB Leipzig        30         23                 15   
4     2017      Schalke 04        29         24                 17   
5     2017      M'gladbach        31         25                 17   
6     2017       Wolfsburg        31         24                 14   
7     2017         FC Koln        24         26                  9   
8     2017      Hoffenheim        31         24                 14   
9     2017          Hertha        26         26                 12   
10    2017           Mainz        31         25                 16   
11    2017         Hamburg        31         25                 15   
12    2017   Werder Bremen        32         26                 19   
13    2017   Ein Frankfurt        34         24                 18   
14    2017        Augsburg        36         26                 20   
15    2017        Freiburg        26         26                  8   
16    2017       Stuttgart        24         24                 16   
17    2017        Hannover        31         25                 13   
18    2016   Bayern Munich        33         25                 18   
19    2016        Dortmund        32         24                 17   
20    2016      Leverkusen        39         23                 21   
21    2016      Schalke 04        39         24                 22   
22    2016       Wolfsburg        43         24                 21   
23    2016      M'gladbach        33         23                 15   
24    2016         Hamburg        38         24                 19   
25    2016           Mainz        39         23                 22   
26    2016         FC Koln        30         24                 13   
27    2016      Hoffenheim        33         24                 15   
28    2016          Hertha        33         24                 14   
29    2016   Werder Bremen        41         25                 19   
..     ...             ...       ...        ...                ...   
438   2006  Greuther Furth        29         24                 14   
439   2006         CZ Jena        37         25                 11   
440   2006     Munich 1860        26         24                 11   
441   2006        RW Essen        27         26                 11   
442   2006      Burghausen        28         25                 13   
443   2006  Erzgebirge Aue        26         27                 11   
444   2006       Offenbach        30         27                  9   
445   2006       Paderborn        28         25                 12   
446   2006    Unterhaching        31         26                 12   
447   2006    Braunschweig        38         26                 14   
448   2006        Augsburg        25         27                  7   
449   2006         Koblenz        28         26                 11   
450   2005     Munich 1860        31         25                 15   
451   2005          Bochum        31         25                 16   
452   2005   Hansa Rostock        31         25                 12   
453   2005        Freiburg        25         24                 14   
454   2005          Aachen        25         27                  6   
455   2005         Dresden        33         26                 13   
456   2005       Karlsruhe        27         25                  7   
457   2005  Greuther Furth        30         25                 14   
458   2005     Saarbrucken        31         26                 15   
459   2005           Ahlen        24         28                  9   
460   2005         Cottbus        26         25                 12   
461   2005  Erzgebirge Aue        25         27                  8   
462   2005      Burghausen        28         25                  9   
463   2005    

# COUNT( )

The first aggregation function we will look at is the __COUNT( )__ function. Previously we have either put a single asterisk in the __SELECT__ statment or we had listed the names of the columns we wanted to be returned. The __COUNT( )__ function can also be listed in the __SELECT__ statement and will count the number of rows specified inbetween the parentheses. The query below has the function __COUNT(*)__ in the __SELECT__ statement. __COUNT(*)__ will count all the rows from the table specified. The below query counts all the rows in the __Teams__ dataset. 

In [9]:
# View the number of rows from Teams dataframe

cur.execute('''SELECT COUNT(*) 
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

COUNT(*)
0       468

The are a total of 468 rows included in the __Teams__ dataset. What happens if we insert a column name instead of an asterisk between the parentheses of the __COUNT( )__ function, for example, __COUNT(Season)__? The function will then count all the rows in the __Season__ column where there is a non-null value. If a value of __Season__ is null for any row, then that row will not be counted. If there are no null values in the __Season__ column then the following function will return 468, the same number as the total number of rows in the __Teams__ dataset.

In [10]:
# Count the number of Season rows in the Teams Dataframe

cur.execute('''SELECT COUNT(Season) 
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

COUNT(Season)
0            468

Great, the number of rows with non-null values is 468, the same as the total amount of rows included in the __Teams__ dataset. That means there are no null values. To practice using the __COUNT( )__ statement, count the number of rows in the __Matches__ dataset and compare your query to the one below:

In [11]:
# Count the number of rows in the Matches dataframe

cur.execute('''SELECT COUNT(*) 
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

COUNT(*)
0     24625

There should be 24,625 rows in the __Matches__ dataset. Check to see if there are any missing values in the column __Season__ from the __Matches__ dataset and compare your query to the one below:

In [12]:
# Count the number of Season rows in the Matches Dataframe

cur.execute('''SELECT COUNT(Season) 
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

COUNT(Season)
0          24625

There are 24,625 rows with non-null values in the __Season__ column in the __Matches__ dataset. Since this is the same as the total number of rows in the __Matches__ dataset, there are no missing/null values in the __Season__ column.

# SELECT DISTINCT

With the __COUNT( )__ function, we were able to see how many rows were listed in both the __Teams__ and __Matches__ datasets. We also saw the number of rows that did not have null values in the __Seasons__ column for both datasets. What if we want to know the unqiue values in the __Seasons__ column in the __Teams__ dataset. We can't use __COUNT(Season)__, because that just gives us the number of rows that have a specified value in the __Season__ column. Instead we will use the __DISTINCT__ clause in the __SELECT__ statement. The below query will show us all of unqiue values that are included in the __Season__ column in the __Teams__ dataset.

In [13]:
# View the unqiue Season values in the Teams Dataframe

cur.execute('''SELECT DISTINCT Season
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Season
0    2017
1    2016
2    2015
3    2014
4    2013
5    2012
6    2011
7    2010
8    2009
9    2008
10   2007
11   2006
12   2005

You can also use more than one column name when using the __DISTINCT__ clause is the __SELECT__ statement. For example, one soccer team may be the home team multiple times in a season, but if we only want to know the seasons each team was a home team we can use the following query:

In [19]:
cur.execute('''SELECT DISTINCT Season, HomeTeam
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Season            HomeTeam
0       2009          Oberhausen
1       2009         Munich 1860
2       2009       Frankfurt FSV
3       2009               Ahlen
4       2009        Union Berlin
5       2009           Paderborn
6       2009           Bielefeld
7       2009      Kaiserslautern
8       2009       Hansa Rostock
9       2009      Greuther Furth
10      2009             Koblenz
11      2009             Cottbus
12      2009            Duisburg
13      2009  Fortuna Dusseldorf
14      2009              Aachen
15      2009            Augsburg
16      2009            St Pauli
17      2009              Bochum
18      2009       Bayern Munich
19      2009          M'gladbach
20      2009               Mainz
21      2009             Hamburg
22      2009           Stuttgart
23      2009            Dortmund
24      2009          Schalke 04
25      2009            Freiburg
26      2009       Werder Bremen
27      2009              Hertha
28      2009            Hannover
29      2009       Ein Frankfurt
...      ...                 ...
1356    2015         Aston Villa
1357    2015         Southampton
1358    2015          Sunderland
1359    2015             Swansea
1360    2015           Tottenham
1361    2015             Watford
1362    2015            West Ham
1363    2015      Crystal Palace
1364    2015            Man City
1365    2015           Liverpool
1366    2016             Burnley
1367    2016      Crystal Palace
1368    2016             Everton
1369    2016                Hull
1370    2016            Man City
1371    2016       Middlesbrough
1372    2016         Southampton
1373    2016             Arsenal
1374    2016         Bournemouth
1375    2016             Chelsea
1376    2016          Man United
1377    2016           Leicester
1378    2016               Stoke
1379    2016             Swansea
1380    2016           Tottenham
1381    2016             Watford
1382    2016           West Brom
1383    2016          Sunderland
1384    2016            West Ham
1385    2016           Liverpool

[1386 rows x 2 columns]

The above results show every distinct __HomeTeam__ that was in every distinct __Season__ from the __Matches__ table. This might seem a bit confusing right now but we will play around more with the __DISTINCT__ clause later. For now, practice selecting all the unique __AwayTeams__ there are in the __Matches__ dataset and compare your query to the one below:

In [22]:
# Show all the unqiue AwayTeams in the Matches Dataframe

cur.execute('''SELECT DISTINCT AwayTeam
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

AwayTeam
0        Kaiserslautern
1             Karlsruhe
2            Leverkusen
3              Nurnberg
4            Schalke 04
5             Stuttgart
6         Werder Bremen
7                Bochum
8              Hannover
9         Hansa Rostock
10               Hertha
11                Mainz
12          Munich 1860
13             St Pauli
14            Wolfsburg
15           M'gladbach
16            Bielefeld
17              Cottbus
18       Greuther Furth
19           Oberhausen
20               Aachen
21                Ahlen
22         Union Berlin
23            Paderborn
24             Augsburg
25              Koblenz
26           Hoffenheim
27        Frankfurt FSV
28   Fortuna Dusseldorf
29        Bayern Munich
..                  ...
98     Sheffield United
99            Liverpool
100         Aston Villa
101              Oldham
102       Nott'm Forest
103           Leicester
104      Crystal Palace
105              Bolton
106       Middlesbrough
107          Sunderland
108               Derby
109            Barnsley
110            Charlton
111            Bradford
112             Watford
113              Fulham
114           West Brom
115         Middlesboro
116          Birmingham
117              Wolves
118          Portsmouth
119               Wigan
120             Reading
121               Stoke
122                Hull
123             Burnley
124           Blackpool
125             Swansea
126             Cardiff
127         Bournemouth

[128 rows x 1 columns]

# Using COUNT( ) & DISTINCT Together

We just saw how to get all the unique __AwayTeam__ names from the Matches dataset above. What happens when we want to know how many unqiue values are returned? to get the count of the unique __AwayTeam__ names, we can use the __COUNT( )__ function together with the __DISTINCT__ clause. The below query will find all the the distinct AwayTeam names in the __Matches__ dataset and return the count or number of rows that are included in distinct AwayTeam names.

In [23]:
# Show all the unqiue AwayTeams in the Matches Dataframe

cur.execute('''SELECT COUNT (DISTINCT AwayTeam)
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

COUNT (DISTINCT AwayTeam)
0                        128

Now we can see all that there are 128 unique Away Teams in the __Matches__ dataset. We can also see that the name of this column is named __COUNT (DISTINCT AwayTeam)__, this is a bit long and unclear. We can change the name of a column using an __AS__ clause. In the __SELECT__ statement, we can rename any column just by writing __AS__ after the data we are selecting and giving it a new name. The below query shows the number of unique away teams in the __Matches__ dataset but this time, the column header is __Num_of_AwayTeams__.

In [25]:
# Show all the unqiue AwayTeams in the Matches Dataframe

cur.execute('''SELECT COUNT (DISTINCT AwayTeam) AS Num_of_AwayTeams
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Num_of_AwayTeams
0               128

To practice using the __COUNT( )__ function and __DISTINCT__ clause together, query the number of distinct __Seasons__ included in the __Teams__ dataset and rename the column to __Num_of_Seasons__. Compare your query to the one below:

In [27]:
# Count of unqiue seasons in the Teams dataset

cur.execute('''SELECT COUNT(DISTINCT Season) AS Num_of_Seasons
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Num_of_Seasons
0              13

# MAX( ) & MIN( ) Functions

We will continue to look at how data can be aggregated in the __SELECT__ statement before we get to the __GROUP BY__ statement. Next we will look at how to use the __MAX( )__ function and __MIN( )__ function. Perhaps we want to know what is the maximum stadium capacity is that is included in the __Teams__ dataset. To do this, we can use the query below:

In [31]:
# Return the maximum StadiumCapacity in the Teams dataset

cur.execute('''SELECT MAX(StadiumCapacity)
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

MAX(StadiumCapacity)
0                81359

The __MAX( )__ function returns the highest or maximum value in a specified column. In the example above, 81,359 was the largest value in the __StadiumCapacity__ column in the __Teams__ dataset. To see what the smallest value is in the __StadiumCapacity__ column, we can use the __MIN( )__ function. The __MIN( )__ function is used to return the minimum value of a column.

In [32]:
# Return the minimum StadiumCapacity in the Teams dataset

cur.execute('''SELECT MIN(StadiumCapacity)
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

MIN(StadiumCapacity)
0                15000

We can see that the smallest stadium capacity included in the __Teams__ dataset is 15,000. To practice using the __MAX( )__ and __MIN( )__ functions, query, the largest number of players in a squad __KaderHome__ and rename the column to __Max_Players__, and the smallest number of players in a squad and rename that column to __Min_Players__ and compare your query to the one below:

In [37]:
# Return the maximum and minimum squad count from the Teams table

cur.execute('''SELECT MAX(KaderHome) AS Max_Players,
                      MIN(KaderHome) AS Min_Players
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Max_Players Min_Players
0          44          23

# AVG( ) Funciton & SUM( ) Function

Now we will look at two more functions that can be used in the __SELECT__ statement. The __AVG( )__ wil return the average value of a specified column. We will continue to look at the __StadiumCapacity__ column in the __Teams__ dataset. The below query shows the average stadium capacity in the __Teams__ dataset.

In [38]:
# Return the average stadium capacity from the Teams table

cur.execute('''SELECT AVG(StadiumCapacity) AS Average_StadiumCapacity
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Average_StadiumCapacity
0             47728.094017

There are a lot of numbers following the decimal. If we want to round to the hundreth place we can add a 
__ROUND( )__ clause around our __AVG( )__ function, like the query below:

In [39]:
# Return the average stadium capacity from the Teams table

cur.execute('''SELECT ROUND(AVG(StadiumCapacity), 2) AS Average_StadiumCapacity
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Average_StadiumCapacity
0                 47728.09

A quick note on the __ROUND( )__ function. The format is as follows:__ROUND(number to be rounded, number of decimal places to round number to)__.
The number we wanted to round was the __AVG(StadiumCapacity)__ and we wanted to round to the nearest hundreth, or __2__ numbers following the decimal, therefore the format we used was __ROUND(AVG(StadiumCapacity), 2)__.
To read more about the __ROUND( )__ functio visit, https://www.w3schools.com/sql/func_sqlserver_round.asp.

The __SUM( )__ function will return the sum of a specified column. For example, if we wanted to see the number of home goals scored in the __Matches__ dataset, we could use the following query:

In [40]:
# Show total number of home goals included in the Matches Dataframe

cur.execute('''SELECT SUM(FTHG) AS Total_HomeGoals
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Total_HomeGoals
0            37357

To practice using the __AVG( )__ and __SUM( )__ functions on your own, query from the __Teams__ dataset the average overall market value of the team pre-season in EUR, __OverallMarketValueHome__, and round the value to the nearest hundreth and then label that column, __Avg_Market_Val__. Also query the sum of the average age of players,  __AvgAgeHome__, and rename that column to __Avg_Age__. Compare your query to the one below:

In [43]:
# Return the average overall market value of the team pre-season and round to the nearest hundreth,
# also return the sum of the average age of players column

cur.execute('''SELECT ROUND(AVG(OverallMarketValueHome), 2) AS Avg_Market_Val,
                      SUM(AvgAgeHome) AS Avg_Age
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Avg_Market_Val  Avg_Age
0     60808525.64    11492

# Bringing Back the WHERE Statement

If you did not read my last blog post about filtering data in the __WHERE__ statement, now may be a good time to take a look, https://medium.com/analytics-vidhya/introduction-to-sql-using-python-filtering-data-with-the-where-statement-80d89688f39e. We are now going to start aggregating filtered data.

What if you wanted to know what how many goals Man United scored during home games on average. To get this answer, we could use the following query:

In [45]:
# Show average number of home goals scored by Man United in the Matches Dataframe

cur.execute('''SELECT ROUND(AVG(FTHG), 2) AS Avg_HomeGoals
               FROM Matches
               WHERE HomeTeam = 'Man United';''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Avg_HomeGoals
0           2.16

By adding the __WHERE__ statement, we are telling our query that we only want to filter data that has a HomeTeam of 'Man United'. To understand this better, we will briefly discuss the SQL order of operations in relation to this query. Although the __SELECT__ statement is the first statement in our SQL query, the first statement to be executed is the __FROM__ statement. Our __FROM__ statment above tells SQL that we are using the __Matches__ dataset as our table. The next statement executed is the __WHERE__ statement. The __WHERE__ statement filters our dataset to include only the information we want, in this case, we only wanted the rows where __HomeTeam = 'Man United'__ to be included. Lastly, the __SELECT__ statement is executed, it returns the data in the form we specify. In this case, we want to return the rounded average value of home goals scored from the data the SQL query filtered through earlier during the query execution. Practice aggregating and using the __WHERE__ statement by writing a query that shows the sum of squad players for all teams during the 2014 seasin from the __Teams__ dataset. Rename the sum of squard players column to __Total_Soccer_Players__. When you are done compare your query to the one below.

In [56]:
# Query the sum of squad players for all teams during the 2014 seasin from the Teams dataset

cur.execute('''SELECT SUM(KaderHome) AS Total_Soccer_Players
               FROM Teams
               WHERE Season = 2014;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Total_Soccer_Players
0                  1164

# GROUP BY STATEMENT & AVG( ) Function

Finally we have reached the __GROUP BY__ function. The __GROUP BY__ function will group together the rows that have matching values in a specified column. To understand how the __GROUP BY__ statement works, we will discuss the query below.

In [60]:
# Return the total home games each team won during the 2016 
# season from the Matches dataset

cur.execute('''SELECT HomeTeam, COUNT(FTR) AS Total_Home_Wins
               FROM Matches
               WHERE FTR = 'H' AND Season = '2016'
               GROUP BY HomeTeam;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

HomeTeam  Total_Home_Wins
0              Arsenal               14
1             Augsburg                5
2        Bayern Munich               13
3            Bielefeld                7
4               Bochum                7
5          Bournemouth                9
6         Braunschweig               13
7              Burnley               10
8              Chelsea               17
9       Crystal Palace                6
10           Darmstadt                6
11            Dortmund               13
12             Dresden                6
13       Ein Frankfurt                7
14      Erzgebirge Aue                7
15             Everton               13
16             FC Koln                9
17  Fortuna Dusseldorf                4
18            Freiburg               10
19      Greuther Furth                9
20             Hamburg                8
21            Hannover               14
22          Heidenheim                7
23              Hertha               12
24          Hoffenheim               11
25                Hull                8
26          Ingolstadt                4
27      Kaiserslautern                8
28           Karlsruhe                3
29           Leicester               10
30          Leverkusen                5
31           Liverpool               12
32          M'gladbach                7
33               Mainz                7
34            Man City               11
35          Man United                8
36       Middlesbrough                4
37         Munich 1860                7
38            Nurnberg                6
39          RB Leipzig               12
40          Sandhausen                6
41          Schalke 04                8
42         Southampton                6
43            St Pauli                5
44               Stoke                7
45           Stuttgart               13
46          Sunderland                3
47             Swansea                8
48           Tottenham               17
49        Union Berlin               11
50             Watford                8
51       Werder Bremen                8
52           West Brom                9
53            West Ham                7
54           Wolfsburg                5
55  Wurzburger Kickers                4

This query returns the total number of home games each team won during the 2016 season. There are two variables in our __SELECT__ statement, __HomeTeam__ and __COUNT(FTR)__. The first variable is an unaggregated column from the __Matches__ dataset, the second variable is an aggregated column from othe __Matches__ dataset. The __WHERE__ statement says to only use data where __FTR = 'H'__ (outcome of a game is a home win) and __Season = 2016__. The __GROUP BY__ groups the rows together by __HomeTeam__. The __GROUP BY__ statement tells the query to combine all the rows with the same home team name together. The result is that each home team is grouped together and every time they won a home game, it was added to the total on the left for the final result. 

In [63]:
# Return the total home games each team won during the 2016 
# season from the Matches dataset

cur.execute('''SELECT HomeTeam, Date, COUNT(FTR) AS Total_Home_Wins
               FROM Matches
               WHERE FTR = 'H' AND Season = '2016'
               GROUP BY HomeTeam;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

HomeTeam        Date  Total_Home_Wins
0              Arsenal  2016-09-10               14
1             Augsburg  2017-04-15                5
2        Bayern Munich  2017-04-08               13
3            Bielefeld  2017-03-17                7
4               Bochum  2017-04-28                7
5          Bournemouth  2016-09-10                9
6         Braunschweig  2017-04-10               13
7              Burnley  2016-08-20               10
8              Chelsea  2016-08-15               17
9       Crystal Palace  2016-09-18                6
10           Darmstadt  2017-02-11                6
11            Dortmund  2016-11-19               13
12             Dresden  2016-10-15                6
13       Ein Frankfurt  2016-11-26                7
14      Erzgebirge Aue  2017-05-14                7
15             Everton  2016-08-27               13
16             FC Koln  2017-04-04                9
17  Fortuna Dusseldorf  2016-08-12                4
18            Freiburg  2016-10-01               10
19      Greuther Furth  2017-04-21                9
20             Hamburg  2017-04-01                8
21            Hannover  2017-01-30               14
22          Heidenheim  2016-09-24                7
23              Hertha  2017-03-11               12
24          Hoffenheim  2017-04-04               11
25                Hull  2016-08-13                8
26          Ingolstadt  2017-01-28                4
27      Kaiserslautern  2016-09-21                8
28           Karlsruhe  2017-03-04                3
29           Leicester  2016-08-27               10
30          Leverkusen  2016-10-01                5
31           Liverpool  2016-09-10               12
32          M'gladbach  2017-02-04                7
33               Mainz  2017-05-13                7
34            Man City  2016-08-13               11
35          Man United  2016-08-19                8
36       Middlesbrough  2016-10-29                4
37         Munich 1860  2016-12-03                7
38            Nurnberg  2016-12-19                6
39          RB Leipzig  2016-09-10               12
40          Sandhausen  2016-09-30                6
41          Schalke 04  2016-11-06                8
42         Southampton  2016-09-18                6
43            St Pauli  2017-02-12                5
44               Stoke  2016-10-15                7
45           Stuttgart  2017-02-26               13
46          Sunderland  2016-11-19                3
47             Swansea  2016-11-26                8
48           Tottenham  2016-08-20               17
49        Union Berlin  2017-04-16               11
50             Watford  2016-09-18                8
51       Werder Bremen  2017-04-16                8
52           West Brom  2016-09-17                9
53            West Ham  2016-08-21                7
54           Wolfsburg  2016-12-17                5
55  Wurzburger Kickers  2016-12-18                4

Try practicing using a __GROUP BY__ statement by writing a query that shows the average number of foreign players for each team from the __Teams__ dataset. Rename the column for average number of foreign players to __Avg_Num_Foreign_Players__ and round the column to the nearest hundreth. Compare your query to the one below:

In [66]:
# View Teams dataframe

cur.execute('''SELECT TeamName, ROUND(AVG(ForeignPlayersHome),2) AS Avg_Num_Foreign_Players
               FROM Teams
               GROUP BY TeamName;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

TeamName  Avg_Num_Foreign_Players
0               Aachen                     8.71
1                Aalen                     7.00
2                Ahlen                     8.67
3             Augsburg                    14.83
4        Bayern Munich                    14.85
5            Bielefeld                    10.20
6               Bochum                    11.92
7         Braunschweig                     9.67
8           Burghausen                    11.00
9              CZ Jena                    13.00
10             Cottbus                    15.33
11           Darmstadt                     9.75
12            Dortmund                    15.15
13             Dresden                     9.67
14            Duisburg                    14.80
15       Ein Frankfurt                    15.85
16      Erzgebirge Aue                     7.60
17             FC Koln                    15.38
18  Fortuna Dusseldorf                    12.22
19       Frankfurt FSV                    14.13
20            Freiburg                    15.31
21      Greuther Furth                    13.38
22             Hamburg                    19.00
23            Hannover                    15.92
24       Hansa Rostock                     9.67
25          Heidenheim                     4.00
26              Hertha                    17.00
27          Hoffenheim                    17.27
28          Ingolstadt                    11.33
29      Kaiserslautern                    14.38
30           Karlsruhe                    10.27
31                Kiel                     3.00
32             Koblenz                    14.50
33          Leverkusen                    14.15
34          M'gladbach                    16.00
35               Mainz                    17.62
36         Munich 1860                    13.33
37            Nurnberg                    15.31
38          Oberhausen                     6.33
39           Offenbach                    10.67
40           Osnabruck                     6.33
41           Paderborn                     9.20
42          RB Leipzig                    14.00
43            RW Essen                    11.00
44          Regensburg                     7.50
45         Saarbrucken                    15.00
46          Sandhausen                     8.33
47          Schalke 04                    17.69
48              Siegen                    13.00
49            St Pauli                     6.82
50           Stuttgart                    15.69
51        Union Berlin                     9.67
52        Unterhaching                    10.50
53               Wehen                    12.50
54       Werder Bremen                    16.38
55           Wolfsburg                    21.54
56  Wurzburger Kickers                    11.00

# ORDER BY Statement

Looking at thhe result we got from the last query we see that the results are organized by __TeamNAme__. What if we wanted to see the teams with the lowest of foreign players at the top of the list and the teams with the highest average of foreign players at the bottom of the list. We can do this by using an __ORDER BY__ statement.

In [67]:
# View Teams dataframe

cur.execute('''SELECT TeamName, ROUND(AVG(ForeignPlayersHome),2) AS Avg_Num_Foreign_Players
               FROM Teams
               GROUP BY TeamName
               ORDER BY ROUND(AVG(ForeignPlayersHome),2);''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

TeamName  Avg_Num_Foreign_Players
0                 Kiel                     3.00
1           Heidenheim                     4.00
2           Oberhausen                     6.33
3            Osnabruck                     6.33
4             St Pauli                     6.82
5                Aalen                     7.00
6           Regensburg                     7.50
7       Erzgebirge Aue                     7.60
8           Sandhausen                     8.33
9                Ahlen                     8.67
10              Aachen                     8.71
11           Paderborn                     9.20
12        Braunschweig                     9.67
13             Dresden                     9.67
14       Hansa Rostock                     9.67
15        Union Berlin                     9.67
16           Darmstadt                     9.75
17           Bielefeld                    10.20
18           Karlsruhe                    10.27
19        Unterhaching                    10.50
20           Offenbach                    10.67
21          Burghausen                    11.00
22            RW Essen                    11.00
23  Wurzburger Kickers                    11.00
24          Ingolstadt                    11.33
25              Bochum                    11.92
26  Fortuna Dusseldorf                    12.22
27               Wehen                    12.50
28             CZ Jena                    13.00
29              Siegen                    13.00
30         Munich 1860                    13.33
31      Greuther Furth                    13.38
32          RB Leipzig                    14.00
33       Frankfurt FSV                    14.13
34          Leverkusen                    14.15
35      Kaiserslautern                    14.38
36             Koblenz                    14.50
37            Duisburg                    14.80
38            Augsburg                    14.83
39       Bayern Munich                    14.85
40         Saarbrucken                    15.00
41            Dortmund                    15.15
42            Freiburg                    15.31
43            Nurnberg                    15.31
44             Cottbus                    15.33
45             FC Koln                    15.38
46           Stuttgart                    15.69
47       Ein Frankfurt                    15.85
48            Hannover                    15.92
49          M'gladbach                    16.00
50       Werder Bremen                    16.38
51              Hertha                    17.00
52          Hoffenheim                    17.27
53               Mainz                    17.62
54          Schalke 04                    17.69
55             Hamburg                    19.00
56           Wolfsburg                    21.54

The query above has an __ORDER BY__ statement. We tell the query to order the results by __ROUND(AVG(ForeignPlayersHome),2)__ and not __Avg_Num_Foreign_Players__. This is because the __SELECT__ statement is executed after the __ORDER BY__ statement, the query does not recognize any variables by there new names as given in the __SELECT__ statement. What if we wanted the highest average number of foreign players to be returned at the top of our results? We can add a __DESC__ clause to the end of the __ORDER BY ROUND(AVG(ForeignPlayersHome, 2)__ statement. See how this works below:

In [71]:
# View Teams dataframe

cur.execute('''SELECT TeamName, ROUND(AVG(ForeignPlayersHome),2) AS Avg_Num_Foreign_Players
               FROM Teams
               GROUP BY TeamName
               ORDER BY ROUND(AVG(ForeignPlayersHome),2) DESC;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

TeamName  Avg_Num_Foreign_Players
0            Wolfsburg                    21.54
1              Hamburg                    19.00
2           Schalke 04                    17.69
3                Mainz                    17.62
4           Hoffenheim                    17.27
5               Hertha                    17.00
6        Werder Bremen                    16.38
7           M'gladbach                    16.00
8             Hannover                    15.92
9        Ein Frankfurt                    15.85
10           Stuttgart                    15.69
11             FC Koln                    15.38
12             Cottbus                    15.33
13            Freiburg                    15.31
14            Nurnberg                    15.31
15            Dortmund                    15.15
16         Saarbrucken                    15.00
17       Bayern Munich                    14.85
18            Augsburg                    14.83
19            Duisburg                    14.80
20             Koblenz                    14.50
21      Kaiserslautern                    14.38
22          Leverkusen                    14.15
23       Frankfurt FSV                    14.13
24          RB Leipzig                    14.00
25      Greuther Furth                    13.38
26         Munich 1860                    13.33
27             CZ Jena                    13.00
28              Siegen                    13.00
29               Wehen                    12.50
30  Fortuna Dusseldorf                    12.22
31              Bochum                    11.92
32          Ingolstadt                    11.33
33          Burghausen                    11.00
34            RW Essen                    11.00
35  Wurzburger Kickers                    11.00
36           Offenbach                    10.67
37        Unterhaching                    10.50
38           Karlsruhe                    10.27
39           Bielefeld                    10.20
40           Darmstadt                     9.75
41        Braunschweig                     9.67
42             Dresden                     9.67
43       Hansa Rostock                     9.67
44        Union Berlin                     9.67
45           Paderborn                     9.20
46              Aachen                     8.71
47               Ahlen                     8.67
48          Sandhausen                     8.33
49      Erzgebirge Aue                     7.60
50          Regensburg                     7.50
51               Aalen                     7.00
52            St Pauli                     6.82
53          Oberhausen                     6.33
54           Osnabruck                     6.33
55          Heidenheim                     4.00
56                Kiel                     3.00

Our results were now returned in DESCENDING order by the average number of foreign players for each team. If we do not put the __DESC__ clause at the end of an __ORDER BY__ statement, the results will be returned in ASCENDING order. You can put __ASC__ at the end of an __ORDER BY__ statement to return results in ascending order but it is not necessary. To practice using the __ORDER BY__ clause, write a query that shows the number of games that resulted in a draw there were for every season. Display the results with the season with the highest number of draws at the top of the results and the season with the lowest number of draws at the bottom of the results. Rename the column with the number of draws to __Num_of_Draws__. Compare your query with the one below:

In [75]:
# Show average number of home goals scored by Man United in the Matches Dataframe

cur.execute('''SELECT Season, COUNT(FTR) AS Num_of_Draws
               FROM Matches
               WHERE FTR = 'D'
               GROUP BY Season
               ORDER BY COUNT(FTR) DESC;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Season  Num_of_Draws
0     1993           340
1     1994           322
2     1998           287
3     1995           286
4     1996           282
5     1997           279
6     2012           276
7     1999           274
8     2014           272
9     2003           267
10    2007           267
11    2015           264
12    2006           263
13    2011           263
14    2008           255
15    2009           254
16    2002           249
17    2000           247
18    2004           246
19    2016           246
20    2010           243
21    2001           241
22    2005           240
23    2013           230

# HAVING Statement

The __HAVING__ statement filters groups. It is always used after the __GROUP BY__ statement and before the __ORDER BY__ statement. It is like the __WHERE__ statement but is only used to filter aggregated rows. To see how it works, we will run the same query we did before but this time we only want to show the seasons where the number of draws was equal to or greater than 275.

In [76]:
# Show average number of home goals scored by Man United in the Matches Dataframe

cur.execute('''SELECT Season, COUNT(FTR) AS Num_of_Draws
               FROM Matches
               WHERE FTR = 'D'
               GROUP BY Season
               HAVING COUNT(FTR) >= 275
               ORDER BY COUNT(FTR) DESC;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Season  Num_of_Draws
0    1993           340
1    1994           322
2    1998           287
3    1995           286
4    1996           282
5    1997           279
6    2012           276

Try practicing using the __HAVING__ statement to query the Sum of Average market value (per player) of the team pre-season in EUR (__AvgMarketValueHome__) for each season where 

In [82]:
# Show average number of home goals scored by Man United in the Matches Dataframe

cur.execute('''SELECT Season, SUM(AvgMarketValueHome) AS Num_of_Draws
               FROM Teams
               GROUP BY Season
               HAVING SUM(AvgMarketValueHome) >= 65000000;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Season  Num_of_Draws
0   2010      65782000
1   2011      69956000
2   2012      73202000
3   2013      75985000
4   2014      84207000
5   2015      87008000
6   2016      84025000
7   2017     100466000

In [ ]:
Average market value (per player) of the team pre-season in EUR

In [ ]:
hen using a __GROUP BY__ statement, every variable in the SELECT statement must either be used to group the data together in the __GROUP BY__ statement, or must be an aggregate function. To see how this works, we wil run the same query, but add another variable.

In [78]:
# View Teams dataframe

cur.execute('''SELECT * 
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Season        TeamName KaderHome AvgAgeHome ForeignPlayersHome  \
0     2017   Bayern Munich        27         26                 15   
1     2017        Dortmund        33         25                 18   
2     2017      Leverkusen        31         24                 15   
3     2017      RB Leipzig        30         23                 15   
4     2017      Schalke 04        29         24                 17   
5     2017      M'gladbach        31         25                 17   
6     2017       Wolfsburg        31         24                 14   
7     2017         FC Koln        24         26                  9   
8     2017      Hoffenheim        31         24                 14   
9     2017          Hertha        26         26                 12   
10    2017           Mainz        31         25                 16   
11    2017         Hamburg        31         25                 15   
12    2017   Werder Bremen        32         26                 19   
13    2017   Ein Frankfurt        34         24                 18   
14    2017        Augsburg        36         26                 20   
15    2017        Freiburg        26         26                  8   
16    2017       Stuttgart        24         24                 16   
17    2017        Hannover        31         25                 13   
18    2016   Bayern Munich        33         25                 18   
19    2016        Dortmund        32         24                 17   
20    2016      Leverkusen        39         23                 21   
21    2016      Schalke 04        39         24                 22   
22    2016       Wolfsburg        43         24                 21   
23    2016      M'gladbach        33         23                 15   
24    2016         Hamburg        38         24                 19   
25    2016           Mainz        39         23                 22   
26    2016         FC Koln        30         24                 13   
27    2016      Hoffenheim        33         24                 15   
28    2016          Hertha        33         24                 14   
29    2016   Werder Bremen        41         25                 19   
..     ...             ...       ...        ...                ...   
438   2006  Greuther Furth        29         24                 14   
439   2006         CZ Jena        37         25                 11   
440   2006     Munich 1860        26         24                 11   
441   2006        RW Essen        27         26                 11   
442   2006      Burghausen        28         25                 13   
443   2006  Erzgebirge Aue        26         27                 11   
444   2006       Offenbach        30         27                  9   
445   2006       Paderborn        28         25                 12   
446   2006    Unterhaching        31         26                 12   
447   2006    Braunschweig        38         26                 14   
448   2006        Augsburg        25         27                  7   
449   2006         Koblenz        28         26                 11   
450   2005     Munich 1860        31         25                 15   
451   2005          Bochum        31         25                 16   
452   2005   Hansa Rostock        31         25                 12   
453   2005        Freiburg        25         24                 14   
454   2005          Aachen        25         27                  6   
455   2005         Dresden        33         26                 13   
456   2005       Karlsruhe        27         25                  7   
457   2005  Greuther Furth        30         25                 14   
458   2005     Saarbrucken        31         26                 15   
459   2005           Ahlen        24         28                  9   
460   2005         Cottbus        26         25                 12   
461   2005  Erzgebirge Aue        25         27                  8   
462   2005      Burghausen        28         25                  9   
463   2005    

# ORDER BY Statement

In [ ]:
# DESC & ASC